libraries/AP_ESC_Telem/AP_ESC_Telem.cpp

```diff
@@ -35,23 +39,327 @@ AP_ESC_Telem::AP_ESC_Telem()
+            motor_freq += rpm * (1.0f / 60.0f);
+            valid_escs++;
+        }
+    }
+    if (valid_escs > 0) {
+        motor_freq /= valid_escs;
+    }
+
+    return motor_freq;
+}
+
+// return all the motor frequencies in Hz for dynamic filtering
+uint8_t AP_ESC_Telem::get_motor_frequencies_hz(uint8_t nfreqs, float* freqs) const
+{
+    uint8_t valid_escs = 0;
+
+    // average the rpm of each motor as reported by BLHeli and convert to Hz
+    for (uint8_t i = 0; i < ESC_TELEM_MAX_ESCS && i < nfreqs; i++) {
+        float rpm;
+        if (get_rpm(i, rpm)) {
+            freqs[valid_escs++] = rpm * (1.0f / 60.0f);
+        }
+    }
+
+    return MIN(valid_escs, nfreqs);
+}
+
+// return number of active ESCs present
+uint8_t AP_ESC_Telem::get_num_active_escs() const {
+    uint8_t nmotors = 0;
+    uint32_t now = AP_HAL::millis();
+    for (uint8_t i = 0; i < ESC_TELEM_MAX_ESCS; i++) {
+        if (now - _telem_data[i].last_update_ms < ESC_TELEM_DATA_TIMEOUT_MS) {
+            nmotors++;
+        }
+    }
+    return nmotors;
+}
+
+// get an individual ESC's slewed rpm if available, returns true on success
+bool AP_ESC_Telem::get_rpm(uint8_t esc_index, float& rpm) const
+{
+    const volatile AP_ESC_Telem_Backend::RpmData& rpmdata = _rpm_data[esc_index];
+
+    if (esc_index > ESC_TELEM_MAX_ESCS || is_zero(rpmdata.update_rate_hz)) {
+        return false;
+    }
+
+    const uint32_t now = AP_HAL::micros();
+    if (rpmdata.last_update_us > 0 && (now - rpmdata.last_update_us < 1000000)) {
+        const float slew = MIN(1.0f, (now - rpmdata.last_update_us) * rpmdata.update_rate_hz * (1.0f / 1e6f));
+        rpm = (rpmdata.prev_rpm + (rpmdata.rpm - rpmdata.prev_rpm) * slew);
+        return true;
+    }
+    return false;
+}
+
+// get an individual ESC's temperature in centi-degrees if available, returns true on success
+bool AP_ESC_Telem::get_temperature(uint8_t esc_index, int16_t& temp) const
+{
+    if (esc_index >= ESC_TELEM_MAX_ESCS
+        || AP_HAL::millis() - _telem_data[esc_index].last_update_ms > ESC_TELEM_DATA_TIMEOUT_MS
+        || !(_telem_data[esc_index].types & AP_ESC_Telem_Backend::TelemetryType::TEMPERATURE)) {
+        return false;
+    }
+    temp = _telem_data[esc_index].temperature_cdeg;
+    return true;
+}
+
+// get an individual motor's temperature in centi-degrees if available, returns true on success
+bool AP_ESC_Telem::get_motor_temperature(uint8_t esc_index, int16_t& temp) const
+{
+    if (esc_index >= ESC_TELEM_MAX_ESCS
+        || AP_HAL::millis() - _telem_data[esc_index].last_update_ms > ESC_TELEM_DATA_TIMEOUT_MS
+        || !(_telem_data[esc_index].types & AP_ESC_Telem_Backend::TelemetryType::MOTOR_TEMPERATURE)) {
+        return false;
+    }
+    temp = _telem_data[esc_index].motor_temp_cdeg;
+    return true;
+}
+
+// get an individual ESC's current in centi-Ampere if available, returns true on success
+bool AP_ESC_Telem::get_current_ca(uint8_t esc_index, uint16_t& amps_ca) const
```

Comment:
make it a float, we do have ESCs that go over 320A, and some can go negative

Reply:
Fixed